# _T. congolense_ Whole Genome Sequencing Project
African animal trypanosomiasis (AAT) is a parasitic disease that causes serious economic losses in livestock from anemia, loss of condition and effects on reproduction. Losses in cattle are especially prominent. It is caused by several parasites of the genus Trypanosoma.

Wildlife is generally resistant to clinical trypanosomiasis in addition to asymptomatic trypanosome infections leading to consider wildlife as 
potential reservoirs for AAT in livestock. Sylvatic transmission is important in AAT epidemiology; however that is yet to be proven. 
This project aims to examine the role of sylvatic transmission of AAT in livestock by quantifying wildlife-to-livestock transmission in 
Shimba Hills National Park by genotyping individual parasites using Whole Genome Sequencing (WGS) data from Illumina sequencing platform.

The main parasite we will be looking in this project is _**Trypanosoma congolense** IL3000_ collected from wild animals, vectors and livestock. The genomic data
and other metadata collected in the field will be used to answer the following questions;
  1. Are trypanosomes transmitted between wildlife and livestock â€“ if so, how often, relative to domestic transmission?
  2. Are there genetic differences between trypanosome strains in wild and domestic animals?
  3. Are there improvements that could be made to management strategies (i.e. of vector, livestock or wildlife) that could reduce AAT burden around the national park?

## Data analysis approach
- Quality check and control of raw reads
- Genome assembly from raw reads 
- Variant calling - GATK pipeline available
- Phylogenetics


## Resources/Tools
- [GATK4 Snakemake variant calling pipeline](https://github.com/LandiMi2/Variant_Calling_Project-)
- Bowtie2 / BWA
